File: 2017/haskell/src/Day04.hs
-- module

module Day04 (day04_1, day04_2) where



-- import

import           Data.List

import           Common



-- solution

day04_1 :: Solution
day04_1 = show . countIf check . map words . lines


day04_2 :: Solution
day04_2 = show . countIf check . map (map sort . words) . lines



-- helpers

check :: Eq a => [a] -> Bool
check []     = True
check (x:xs) = notElem x xs && check xs
-- check l = l == nub l

File: 2017/rust/src/day_04.rs
// solutions

pub fn part1(input: &str) -> String {
    let mut res: i32 = 0;
    for line in input.lines() {
        let words: Vec<&str> = line.split_whitespace().collect();
        if check(&words) {
            res += 1
        }
    }
    return format!("{}", res);
}

pub fn part2(input: &str) -> String {
    let mut res: i32 = 0;
    for line in input.lines() {
        let words: Vec<Vec<char>> = line.split_whitespace().map(sorted_chars).collect();
        if check(&words) {
            res += 1
        }
    }
    return format!("{}", res);
}



// helpers

fn check<T: PartialEq>(words: &Vec<T>) -> bool {
    for i in 0 .. words.len() {
        for j in 0 .. i {
            if words[i] == words[j] {
                return false;
            }
        }
    }
    return true;
}

fn sorted_chars(word: &str) -> Vec<char> {
    let mut res: Vec<char> = word.chars().collect();
    res.sort();
    return res;
}
